File: src/ruby/fibonacci_memoization.rb
def fibonacci(n, memo = {})
  if n <= 1
    return 1
  end

  # Check if the result is already memoized
  if memo.key?(n)
    return memo[n]
  end

  # Calculate Fibonacci recursively and store the result in memoization
  result = fibonacci(n - 1, memo) + fibonacci(n - 2, memo)
  memo[n] = result

  return result
end

n = 16
puts "Fib(#{n}): #{fibonacci(n)}"

File: src/go/fibonacci_memoization.go
package main

import "fmt"

var memo = make(map[int]int)

func Fibonacci(n int) int {
	if n <= 1 {
		return 1
	}

	// Check if the value is already memoized
	if val, ok := memo[n]; ok {
		return val
	}

	// Calculate Fibonacci recursively and store the result in memoization
	result := Fibonacci(n-1) + Fibonacci(n-2)
	memo[n] = result
	return result
}

func main() {
	n := 9
	fmt.Println("Fibonacci Memoization:", Fibonacci(n))
}
